feat: Add SASL/OAUTHTOKEN support

einarmo opened the pull request:
Closes #252.

This adds `SASL/OAUTHBEARER` support to the client.

Since `SaslConfig` no longer always contains `Credentials` I did some minor refactoring and moved the rsasl `SASLConfig` building into a separate method on `SaslConfig`. There is no convenient method for doing Oauth, so we have to implement an rsasl callback for providing a bearer token.

I've chosen the token callback to be asynchronous, since users of this library are likely to use an asynchronous HTTP library, or other OAUTH mechanism, and mixing sync and async is a very bad idea.

One thing I have not done is write any tests, mostly since writing an actual integration test supporting this would be very complicated. Since I have now done some of that myself I have a vague idea of what would be needed:

 - You would need an IDP, we used Azure AD, but you could make that self contained using Keycloak or similar, though configuring that as part of tests is also non-trivial.
 - You would need some form of Oauth plugin, which I believe isn't part of the docker image you use in tests (not entirely sure about that, we use the `strimzi` image).

- [x] I've read the contributing section of the project [CONTRIBUTING.md](https://github.com/influxdata/rskafka/blob/main/CONTRIBUTING.md).
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed).

einarmo commented:
Looks like audit is failing due to rust version. A simple fix would be to lock `cargo-deny` to `0.16.2`.

crepererum commented on Cargo.toml:
```diff
@@ -30,13 +30,14 @@ lz4 = { version = "1.23", optional = true }
 parking_lot = "0.12"
 rand = "0.8"
 rustls = { version = "0.23", optional = true, default-features = false, features = ["logging", "ring", "std", "tls12"] }
+serde = "^1.0.210"
```
This doesn't seem to be used anywhere?

einarmo replied:
There's a version conflict introduced by strict bounds on `serde` in `rsasl` and `criterion` (I think), this is needed for version resolution.

crepererum replied:
I'll have a look at that :eyes:

crepererum replied:
With this line removed, I have zero issues. Could you try to remove your `Cargo.lock` and see if it works w/o this line? If not, what exact issue are you facing (i.e. can you post the error message)?

einarmo replied:
Huh. You're right. My bad. I did get some weird error but removing Cargo.lock worked.

crepererum commented on src/protocol/frame.rs:
```diff
@@ -163,7 +163,7 @@ mod tests {
         data.set_position(0);
         let actual = data.read_message(0).await.unwrap();
-        assert_eq!(actual, vec![]);
+        assert!(actual.is_empty())
```
why was that changed?

einarmo replied:
I have no clue why, but on newer rust versions the old approach fails to compile due to type resolution.

crepererum replied:
OK, then that's fine :slightly_smiling_face:

crepererum commented on src/protocol/frame.rs:
```diff
@@ -172,6 +172,6 @@ mod tests {
 
         let actual = server.read_message(0).await.unwrap();
-        assert_eq!(actual, vec![]);
+        assert!(actual.is_empty())
```
same question as above

crepererum replied:
See https://github.com/influxdata/rskafka/pull/253#discussion_r1901708016

crepererum commented on src/messenger.rs:
```diff
@@ -205,6 +201,12 @@ pub enum SaslError {
+    #[error("Invalid SASL config: {0}")]
+    InvalidConfig(#[from] SASLError),
+
+    #[error("Other SASL error: {0}")]
+    Other(String),
```
I guess you've introduced this variant so that you have a catch-all version for the callback? I'm wondering if the callback should even use `SaslError` in the first place or if it should rather return `Box<dyn std::error::Error + Send + Sync>` and you convert this to a `SaslError` variant like

```rust
#[derive(Error, Debug)]
pub enum SaslError {
    // ...

    #[error("OAuth Callback: {0}")]
    OauthCallback(Box<dyn std::error::Error + Send + Sync>),

    // ...
}
```

einarmo replied:
That's probably a good idea. I'll change it :+1:

crepererum commented on src/connection/transport/sasl.rs:
```diff
@@ -30,19 +46,100 @@ impl Credentials {
+            )?),
+            Self::Oauthbearer(credentials) => {
+                // Fetch the token first, since that's an async call.
+                // The user can and should cache the token as appropriate to their OAUTH provider.
```
```rust
// The user can and should cache the token as appropriate to their OAUTH provider.
```

Could you move this particular note into the docstring for `OauthCallback`? I think this is important for the user.

crepererum reviewed (changes requested):
Looks pretty good for the first round. Couple of smaller nitpicks. I've fixed the CI in #254, so if you rebase this branch, you should be fine.

einarmo commented:
Just FYI: I added a second commit, since I ran into a bug in the existing SASL implementation. It seems like `rsasl` sometimes (for PLAIN auth, among others), expects us to send a final message even when `state` is marked as finished.

I changed the code there a bit to account for this, I really don't understand how this could work at all, it should have completely broken PLAIN auth.

crepererum commented:
> I changed the code there a bit to account for this, I really don't understand how this could work at all, it should have completely broken PLAIN auth.

me neither, but if it works now it works :shrug:

crepererum reviewed (approved):
thank you
